Assets/MRTK/Providers/WindowsMixedReality/XRSDK/WindowsMixedRealityDeviceManager.cs

```diff
@@ -126,9 +144,99 @@ public override bool CheckCapability(MixedRealityCapability capability)
             return false;
         }
 
+#if HP_CONTROLLER_ENABLED
+        private MotionControllerWatcher motionControllerWatcher;
+
+        /// <summary>
+        /// Dictionary to capture all active HP controllers detected
+        /// </summary>
+        private readonly Dictionary<uint, MotionControllerState> trackedMotionControllerStates = new Dictionary<uint, MotionControllerState>();
+
+        private readonly Dictionary<uint, GenericXRSDKController> activeMotionControllers = new Dictionary<uint, GenericXRSDKController>();
+#endif
+
         #endregion IMixedRealityCapabilityCheck Implementation
 
+
         #region Controller Utilities
+#if HP_CONTROLLER_ENABLED
+        private static readonly ProfilerMarker GetOrAddControllerPerfMarker = new ProfilerMarker("[MRTK] WindwosMixedRealityXRSDKDeviceManager.GetOrAddController");
+
+        protected override GenericXRSDKController GetOrAddController(InputDevice inputDevice)
+        {
+            using (GetOrAddControllerPerfMarker.Auto())
+            {
+                GenericXRSDKController detectedController = base.GetOrAddController(inputDevice);
+                SupportedControllerType currentControllerType = GetCurrentControllerType(inputDevice);
+
+                // Add the Motion Controller state if it's an HPMotionController
+                uint controllerId = GetControllerId(inputDevice);
+                if (currentControllerType == SupportedControllerType.HPMotionController)
+                {
+                    ((HPMotionController)detectedController).MotionControllerState = trackedMotionControllerStates[controllerId];
```

Comment:
In the other device manager, this checks to make sure the ID is found in `trackedMotionControllerStates` before trying to access it. Could lead to a KeyNotFoundException